servers/audio/audio_server.cpp

```diff
@@ -1839,9 +1842,75 @@ String AudioServer::get_input_device() {
 }
 
 void AudioServer::set_input_device(const String &p_name) {
+	// how about: if (!p_is_active) here?
 	AudioDriver::get_singleton()->set_input_device(p_name);
 }
 
+Error AudioServer::set_input_device_active(bool p_is_active) {
+	if (input_device_active == p_is_active) {
+		return OK;
+	}
+	if (p_is_active) {
+		if (!GLOBAL_GET("audio/driver/enable_input")) {
+			WARN_PRINT("You must enable the project setting \"audio/driver/enable_input\" to use audio capture.");
+			return FAILED;
+		}
+
+		input_buffer_ofs = 0;
+		input_device_active = true;
+		return AudioDriver::get_singleton()->input_start();
+	}
+	input_device_active = false;
+	return AudioDriver::get_singleton()->input_stop();
+}
+
+int AudioServer::get_input_frames_available() {
+	AudioDriver *ad = AudioDriver::get_singleton();
+	ad->lock();
+	int input_position = ad->get_input_position();
+	if (input_position < input_buffer_ofs) {
+		int buffsize = ad->get_input_buffer().size();
+		input_position += buffsize;
+	}
+	ad->unlock();
+	return (input_position - input_buffer_ofs) / 2;
+}
+
+int AudioServer::get_input_buffer_length_frames() {
+	AudioDriver *ad = AudioDriver::get_singleton();
+	ad->lock();
+	int buffsize = ad->get_input_buffer().size();
+	ad->unlock();
+	return buffsize / 2;
+}
+
+PackedVector2Array AudioServer::get_input_frames(int p_frames) {
+	PackedVector2Array ret;
+	AudioDriver *ad = AudioDriver::get_singleton();
+	ad->lock();
+	int input_position = ad->get_input_position();
+	Vector<int32_t> buf = ad->get_input_buffer(); // should this be returning Vector<int32_t>& ??
```

Comment:
> // should this be returning Vector<int32_t>& ??

Yes! Or rather, it should return `const Vector<int32_t>&`. Feel free to adjust accordingly.

Reply:
Originally I did use `Vector<int32_t>&`, but  @lyuma was adamant that this made things worse because the Vector class contains a CowData array, so it's not being copied if nothing is written. https://github.com/godotengine/godot/blob/7ed0b6167618f9c8c5df87c7e83da6003c081368/core/templates/cowdata.h